docs/proposals/node-allocatable.md

```diff
@@ -0,0 +1,155 @@
+   workloads. For example, processes scheduled to run alongside Kubernetes on Mesos.
+3. **Kubelet Allocatable** - Compute resources available for scheduling (including scheduled &
+   unscheduled resources). This value is currently implicit, and the focus of this proposal. See
+   [below](#api-changes) for more details.
+4. **System Reserved** (proposed) - Compute resources reserved for system components such as the docker daemon, kubelet, kube proxy, etc.
+
+### API changes
+
+#### Allocatable
+
+Add `Allocatable` (4) to
+[`NodeStatus`](https://htmlpreview.github.io/?https://github.com/kubernetes/kubernetes/HEAD/docs/api-reference/v1/definitions.html#_v1_nodestatus):
+
+```
+type NodeStatus struct {
+  ...
+  // Allocatable represents schedulable resources of a node.
+  Allocatable ResourceList `json:"allocatable,omitempty"`
+  ...
+}
+```
+
+Allocatable will be computed by the Kubelet and reported periodically to the API server. It is
+defined to be:
+
+```
+   [Allocatable] = [Machine Capacity] - ( [Reserved] + [System Reserved] + [Kernel usage (smoothed)] )
+```
+
+The scheduler will use `Allocatable` in place of `Capacity` when scheduling pods, and the Kubelet
+will use it when perming admission checks.
+
+- Since kernel usage can fluctuate and is out of kubernetes control, `Allocatable` may vary over
+  time.
+- Scheduler can compute unscheduled resources as `[Allocatable] - [Usage]`, where usage is gathered
+  from the [kubelet metrics API](compute-resource-metrics-api.md).
+
+#### System Reserved
+
+- **TODO**: Determine which resources *must* be specified (CPU+memory?)
+- **TODO**: Decide on format (parsed string vs. individual resource flags)
+
+`SystemReserved` is a **required** parameter specifying resources reserved for system components
+(4). It is provided as a command-line flag to the Kubelet at startup, and therefore cannot be
+changed during normal Kubelet operation (this could change in the [future](#future-work)).
+
+#### Reserved
+
+`Reserved` is an **optional** parameter specifying reserved resources (2). It is specified as a
+Kubelet command line flag using the same format as [system reserved](#system-reserved).
+
+## Issues
+
+### System reservation too small
+
+Solution **TBD**
+
+**Option 1:** Do nothing (best effort). Let the system daemons overflow the reserved resources and
+  hope for the best. If the machine usage is less than Allocatable, there will be some room for
+  overflow and the node should continue to function. If the machine has been scheduled to capacity
+  (worst-case scenario), it may enter an unstable state.
+
+**Option 2:** Assume the worst and transition the node to the not-ready state. The kubelet could
+  attempt to reduce usage by evicting pods until usage comes under control. At that point the node
+  could be made available again.
+
+### Version discrepency
+
+**API server / scheduler is not allocatable-resources aware:** The kubelet already performs
```

Comment:
> Interesting. For my own understanding, why does the scheduler retry when the node rejects a pod? Is there a back off? Does the scheduler ever give up (cluster is full)?

Yeah, today the scheduler doesn't have any backoff, and never gives up. Your implied observation is correct, that such features would mitigate this problem. But since we need to consider user-created schedulers (which may not do these things) anyway, I think we need to use the staged rollout approach.

> The 2 stage rollout is a good idea. I think we can start using Allocatable in the scheduler when it is reported by the Kubelet (which should cover the normal case) in the first stage, and only postpone Kubelet admission checks to the second stage (as you suggested). Is that correct?

Yes.

> Would we space the stages across versions (e.g. Kubelet reports & schedular uses it in 1.2, and Kubelet uses it in 1.3)?

Yeah, I think something like that is right. @bgrant0607 is the expert on how we change and deprecate system behavior, though. :)